File: src/core/gap-analysis.js
/**
 * Calculate gaps between current and target identity capabilities.
 * Returns sorted gaps with most severe first.
 */
export function calculateGap(requirements) {
  const gaps = (requirements || []).map((req) => {
    const gap = Math.max(req.targetLevel - (req.currentLevel ?? 0), 0);
    return {
      domain: req.domain,
      capability: req.capability,
      targetLevel: req.targetLevel,
      currentLevel: req.currentLevel ?? 0,
      gap,
      rationale: req.rationale
    };
  });

  return gaps.sort((a, b) => b.gap - a.gap || b.targetLevel - a.targetLevel);
}

export function computeCapabilityGaps(identityState = [], requirements = []) {
  const stateIndex = new Map(
    (identityState || []).map((entry) => [
      key(entry.domain, entry.capability),
      { ...entry, level: clamp(entry.level, 1, 10) }
    ])
  );

  return (requirements || []).map((req) => {
    const currentEntry = stateIndex.get(key(req.domain, req.capability));
    const currentLevel = currentEntry ? currentEntry.level : 3;
    const rawGap = Math.max((req.targetLevel || 0) - currentLevel, 0);
    const weight = clamp(req.weight ?? 0, 0, 1);
    const weightedGap = rawGap * weight;
    return {
      requirementId: req.id,
      domain: req.domain,
      capability: req.capability,
      targetLevel: req.targetLevel,
      currentLevel,
      weight,
      rawGap,
      weightedGap
    };
  });
}

export function rankCapabilityGaps(gaps = []) {
  const ranked = [...(gaps || [])].sort((a, b) => {
    if (b.weightedGap !== a.weightedGap) return b.weightedGap - a.weightedGap;
    return String(a.capability).localeCompare(String(b.capability));
  });
  return ranked.map((gap, idx) => ({ ...gap, rank: idx + 1 }));
}

function key(domain, capability) {
  return `${domain || ''}:${capability || ''}`.toLowerCase();
}

function clamp(val, min, max) {
  const num = Number(val);
  if (Number.isNaN(num)) return min;
  return Math.min(Math.max(num, min), max);
}

File: backend/jyriko/domain/gap_analysis.py
"""Capability gap analysis — port of src/core/gap-analysis.js."""
from typing import Any


def _key(domain: Any, capability: Any) -> str:
    return f"{domain or ''}:{capability or ''}".lower()


def _clamp(val: Any, lo: float, hi: float) -> float:
    try:
        return min(max(float(val), lo), hi)
    except (TypeError, ValueError):
        return lo


def calculate_gap(requirements: list[dict[str, Any]] | None = None) -> list[dict[str, Any]]:
    requirements = requirements or []
    gaps = [
        {
            "domain": req.get("domain"),
            "capability": req.get("capability"),
            "targetLevel": req.get("targetLevel"),
            "currentLevel": req.get("currentLevel") or 0,
            "gap": max((req.get("targetLevel") or 0) - (req.get("currentLevel") or 0), 0),
            "rationale": req.get("rationale"),
        }
        for req in requirements
    ]
    return sorted(gaps, key=lambda g: (-g["gap"], -(g.get("targetLevel") or 0)))


def compute_capability_gaps(
    identity_state: list[dict[str, Any]] | None = None,
    requirements: list[dict[str, Any]] | None = None,
) -> list[dict[str, Any]]:
    identity_state = identity_state or []
    requirements = requirements or []
    state_index: dict[str, dict[str, Any]] = {
        _key(e.get("domain"), e.get("capability")): {**e, "level": _clamp(e.get("level"), 1, 10)}
        for e in identity_state
    }

    def _current(req: dict[str, Any]) -> float:
        k = _key(req.get("domain"), req.get("capability"))
        return state_index[k]["level"] if k in state_index else 3.0

    return [
        {
            "requirementId": req.get("id"),
            "domain": req.get("domain"),
            "capability": req.get("capability"),
            "targetLevel": req.get("targetLevel"),
            "currentLevel": _current(req),
            "weight": _clamp(req.get("weight") or 0, 0, 1),
            "rawGap": max((req.get("targetLevel") or 0) - _current(req), 0),
            "weightedGap": _clamp(req.get("weight") or 0, 0, 1) * max(
                (req.get("targetLevel") or 0) - _current(req), 0
            ),
        }
        for req in requirements
    ]


def rank_capability_gaps(gaps: list[dict[str, Any]] | None = None) -> list[dict[str, Any]]:
    gaps = gaps or []
    ranked = sorted(gaps, key=lambda g: (-(g.get("weightedGap") or 0), str(g.get("capability") or "")))
    return [{**g, "rank": idx + 1} for idx, g in enumerate(ranked)]
